### PR TITLE
Add Eventing 

### PR DESCRIPTION
**Pull request by skothari-tibco**
**What kind of change does this PR introduce?** (check one with "x")
```
[] Bugfix
[x] Feature
[] Code style update (formatting, local variables)
[] Refactoring (no functional changes, no api changes)
[] Other... Please describe:
```

**Fixes**: #

**What is the current behavior?**

**What is the new behavior?**
Adding event listeners support in core . PR TIBCOSoftware/flogo-lib#241

### CODE REVIEW

File: support/event/events.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	"sync"
+
+	"github.com/project-flogo/core/support/log"
+)
+
+const (
+	ENV_PUBLISH_AUDIT_EVENTS_KEY = "FLOGO_PUBLISH_AUDIT_EVENTS"
+)
+
+type EventListener interface {
```
[the remark this answers is not in the document]
Comment (skothari-tibco) [reply support/event/events.go]: Yup, makes sense.

[the remark this answers is not in the document]
Comment (debovema) [reply support/event/events.go]: I'm wondering about the need to put event-related packages as children of "support".

[the remark this answers is not in the document]
Comment (fm-tibco) [reply support/event/events.go]: There isn't much to the event package, it can be just event.go, right in support, no need for a package.

[the remark this answers is not in the document]
Comment (fm-tibco) [reply support/event/events.go]: @vijaynalawade @mellistibco thoughts?

[the remark this answers is not in the document]
Comment (fm-tibco) [reply support/event/events.go]: ignoring whether the amount of code warrants a package.  We can also consider the aesthetics:
```
event.Listener
event.RegisterListener
```
vs 
```
support.EventListener
support.RegisterEventListener
```

[the remark this answers is not in the document]
Comment (vijaynalawade) [reply support/event/events.go]: @fm-tibco : I would prefer event.Listener/event.RegisterListener as app eventing is first class in OSS. In fact, we should make it more verbose like `appevent` etc.

[the remark this answers is not in the document]
Comment (debovema) [reply support/event/events.go]: +1 with @vijaynalawade : I think it should be an *event* package and it is already coded like that.

My main concern is about the import path. I don't understand why there is a *support* directory leading to import paths such as *github.com/project-flogo/core/support/log* or **github.com/project-flogo/core/support/event*.

Such packages should either live in their own path or IMHO in a subpath of the *engine* (*github.com/project-flogo/core/engine/event* for instance).

[the remark this answers is not in the document]
Comment (skothari-tibco) [reply support/event/events.go]: @debovema the reason I placed it in support is because I believe it's a part of non functional feature of core. But I would like to know what @fm-tibco @mellistibco and @vijaynalawade  think about it ?

[the remark this answers is not in the document]
Comment (fm-tibco) [reply support/event/events.go]: I don't believe log or event should be top level packages, they function in a sort of "support" capacity at this moment and is not really core functionality/construct.  As for the packages living under engine, I'd rather not have an activity/trigger require a dependency on the engine package just for logging.  Maybe an alternative package name?

[the remark this answers is not in the document]
Comment (debovema) [reply support/event/events.go]: @fm-tibco : I am not shocked of *log* being a top level package. If there is the possibility of this package being used without other base features of the *core* repository, it could even be moved to a separate Go module / Git repository. But it surely is a little overkill :)
For the *event* package, I think it can live in the *engine* because even though there is no compile dependency between packages using events and the *engine* package, on a runtime point-of-view they are bound.

[the remark this answers is not in the document]
Comment (fm-tibco) [reply support/event/events.go]: @debovema, you make a good point, so I'm fine with moving events under engine and leaving log under support.  @vijaynalawade, @mellistibco thoughts?

[the remark this answers is not in the document]
Comment (mellistibco) [reply support/event/events.go]: +1 i like having events under the engine package, seems to make a lot of sense...

File: engine/event/events.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+	if len(eventListeners) == 0 {
+		// No more listeners. Stop go routine
+		shutdown <- true
+		publisherRoutineStarted = false
+	}
+}
+
+//  EventContext is a wrapper over specific event
+type EventContext struct {
```
Comment (fm-tibco) [inline engine/event/events.go]: this can probably just be Context

File: examples/activity/activity.go [hunk 3]
```diff
@@ -42,15 +42,15 @@ func (a *Activity) Eval(ctx activity.Context) (done bool, err error) {
 	input := &Input{}
 	err = ctx.GetInputObject(input)
 	if err != nil {
-		return false, err
+		return true, err
```
Comment (fm-tibco) [inline examples/activity/activity.go]: Why was this changed from false to true?

Comment (skothari-tibco) [reply examples/activity/activity.go]: Wouldn't we want to exit the activity when we encounter error?

Comment (fm-tibco) [reply examples/activity/activity.go]: If there is an error, that first parameter gets ignored, so it's returning the go default value for a boolean essentially.  Kind of like when you return nil, err in a lot of cases

Comment (skothari-tibco) [reply examples/activity/activity.go]: Alright. In the old code it had to be true as far as I know and that's why I changed it to true.
